File: include/nole.php
<?php
namespace Nole;
use Nol\Document as NolDocument;
use Nol\Builder as NolBuilder;

class Document extends NolDocument {}
class Builder extends NolBuilder {}

class Parser {
    private $text, $pos = 0, $start, $depth = 0, $nole;
    public function __construct($text, $nole = false) { $this->text = $text; $this->start = microtime(true); $this->nole = $nole; }
    private function peek() { return $this->pos < strlen($this->text) ? $this->text[$this->pos] : null; }
    private function advance() { return $this->pos < strlen($this->text) ? $this->text[$this->pos++] : null; }
    private function skipWs() {
        while ($this->pos < strlen($this->text)) {
            if (ctype_space($this->text[$this->pos])) $this->advance();
            elseif ($this->text[$this->pos] === '#') {
                $this->advance();
                if ($this->peek() === '#') {
                    $this->advance();
                    while ($this->pos < strlen($this->text) - 1 && substr($this->text, $this->pos, 2) !== '##') $this->advance();
                    if ($this->pos < strlen($this->text)) { $this->advance(); $this->advance(); }
                } else {
                    while ($this->pos < strlen($this->text) && $this->text[$this->pos] !== "\n") $this->advance();
                }
            } else break;
        }
    }
    private function readStr($q) {
        $s = "";
        while ($this->pos < strlen($this->text)) {
            $c = $this->advance();
            if ($c === $q) return $s;
            if ($c === '\\') {
                $e = $this->advance();
                if ($e === 'u') { $s .= mb_chr(hexdec(substr($this->text, $this->pos, 4))); $this->pos += 4; }
                elseif ($e === 'U') { $s .= mb_chr(hexdec(substr($this->text, $this->pos, 8))); $this->pos += 8; }
                else { $s .= ['n'=>"\n", 'r'=>"\r", 't'=>"\t"][$e] ?? $e; }
            } else $s .= $c;
        }
        return $s;
    }
    private function readKey() {
        $this->skipWs(); $c = $this->peek();
        if ($c === '"' || $c === "'") return $this->readStr($this->advance());
        $k = "";
        while ($this->pos < strlen($this->text) && (ctype_alnum($this->text[$this->pos]) || in_array($this->text[$this->pos], ['_', '-']))) $k .= $this->advance();
        return $k;
    }
    private function parseValue() {
        $this->skipWs();
        if (microtime(true) - $this->start > 1) throw new \Exception("Timeout");
        if (++$this->depth > 100) throw new \Exception("Depth");
        $c = $this->peek();
        if ($c === '{') {
            $this->advance(); $o = [];
            while ($this->peek() && $this->peek() !== '}') {
                $this->skipWs(); $this->parsePair($o);
                $this->skipWs(); if ($this->peek() === ',') $this->advance();
            }
            if ($this->peek() === '}') $this->advance();
            $res = $o;
        } elseif ($c === '[') {
            $this->advance(); $a = [];
            while ($this->peek() && $this->peek() !== ']') {
                $this->skipWs(); $a[] = $this->parseValue();
                $this->skipWs(); if ($this->peek() === ',') $this->advance();
            }
            if ($this->peek() === ']') $this->advance();
            $res = $a;
        } elseif ($this->nole && ($c === '*' || $c === '<')) {
            $t = $this->advance();
            if ($t === '<') {
                $b = ""; while ($this->peek() && $this->peek() !== '>') $b .= $this->advance();
                if ($this->peek() === '>') $this->advance();
                $res = ["_coerce" => ["type" => $b, "value" => $this->parseValue()]];
            } else {
                $b = ""; while ($this->peek() && (ctype_alnum($this->peek()) || in_array($this->peek(), ['_', '-']))) $b .= $this->advance();
                $res = "*" . $b;
            }
        } elseif ($c === '"' || $c === "'") $res = $this->readStr($this->advance());
        elseif ($c && (ctype_digit($c) || $c === '-')) {
            $s = "";
            while ($this->pos < strlen($this->text) && (ctype_digit($this->text[$this->pos]) || strpos(".-eE+", $this->text[$this->pos]) !== false)) $s .= $this->advance();
            $res = (strpos($s, '.') !== false || stripos($s, 'e') !== false) ? floatval($s) : intval($s);
        } else {
            $s = "";
            while ($this->pos < strlen($this->text) && ctype_alpha($this->text[$this->pos])) $s .= $this->advance();
            if ($s === "true") $res = true;
            elseif ($s === "false") $res = false;
            elseif ($s === "null") $res = null;
            else throw new \Exception("Invalid value: $s");
        }
        $this->depth--; return $res;
    }
    private function parsePair(&$o) {
        $this->skipWs();
        if ($this->nole && $this->peek() === '&') {
            $this->advance(); $n = $this->readKey(); $this->skipWs();
            $val = ($this->peek() === ':' && ($this->advance())) ? $this->parseValue() : null;
            if (!isset($o["_anchors"])) $o["_anchors"] = [];
            $o["_anchors"][] = ["name" => $n, "value" => $val]; return;
        }
        $is_m = false; if ($this->nole && $this->peek() === '<') { $this->advance(); if ($this->peek() === '<') { $this->advance(); $is_m = true; } else $this->pos--; }
        $k = $is_m ? "<<" : $this->readKey();
        $this->skipWs(); if ($this->peek() === ':') $this->advance();
        $v = $this->parseValue();
        if ($is_m) { if (!isset($o["<<"])) $o["<<"] = []; $o["<<"][] = $v; }
        else { if (array_key_exists($k, $o)) throw new \Exception("Dup: $k"); $o[$k] = $v; }
    }
    public function parse() {
        $root = [];
        while ($this->pos < strlen($this->text)) {
            $this->skipWs(); if (!$this->peek()) break;
            if ($this->peek() === '[') {
                $this->advance(); $is_a = false; if ($this->peek() === '*') { $this->advance(); $is_a = true; }
                $path = ""; while ($this->peek() && $this->peek() !== ']') $path .= $this->advance();
                if ($this->peek() === ']') $this->advance();
                $parts = explode('.', $path); $curr = &$root;
                foreach (array_slice($parts, 0, -1) as $p) {
                    if (!isset($curr[$p]) || !is_array($curr[$p])) $curr[$p] = [];
                    $curr = &$curr[$p];
                }
                $last = end($parts);
                if ($is_a) {
                    if (!isset($curr[$last]) || !is_array($curr[$last])) $curr[$last] = [];
                    $entry = []; $curr[$last][] = &$entry; $this->parseInto($entry);
                } else {
                    if (!isset($curr[$last]) || !is_array($curr[$last])) $curr[$last] = [];
                    $this->parseInto($curr[$last]);
                }
            } else $this->parsePair($root);
        }
        return $root;
    }
    private function parseInto(&$o) {
        while ($this->pos < strlen($this->text)) {
            $this->skipWs(); if ($this->peek() === '[' || !$this->peek()) break;
            $this->parsePair($o);
        }
    }
}

class Evaluator {
    private $anchors = [], $app_env, $doc_env = [];
    public function __construct($app_env) { $this->app_env = array_flip($app_env); }
    public function evaluate($root) {
        $root = $this->collectMeta($root);
        $root = $this->resolveMerges($root);
        $root = $this->resolveEnv($root);
        $root = $this->resolveInterp($root, $root);
        return new Document($this->resolveCoerce($root));
    }
    private function dump($v) {
        if (is_null($v)) return "null"; if (is_bool($v)) return $v ? "true" : "false";
        if (is_numeric($v)) { if (is_int($v)) return strval($v); $s = sprintf("%.5f", $v); while (str_ends_with($s, '0')) $s = substr($s, 0, -1); if (str_ends_with($s, '.')) $s .= '0'; return $s; }
        if (is_string($v)) return $v;
        if (array_is_list($v)) return "[" . implode(", ", array_map([$this, 'dump'], $v)) . "]";
        $items = []; foreach ($v as $k => $val) if ($k[0] !== '_') $items[] = "$k: " . $this->dump($val);
        return "{" . implode(", ", $items) . "}";
    }
    private function collectMeta($v) {
        if (is_array($v) && !array_is_list($v)) {
            if (isset($v["_anchors"])) { foreach ($v["_anchors"] as $a) { $this->anchors[$a["name"]] = $this->collectMeta($a["value"] === null ? $v : $a["value"]); } unset($v["_anchors"]); }
            if (isset($v["_env"]) && isset($v["_env"]["allowed"])) { foreach ($v["_env"]["allowed"] as $x) $this->doc_env[$x] = true; }
            $res = []; foreach ($v as $k => $val) $res[$k] = $this->collectMeta($val); return $res;
        }
        if (is_array($v)) return array_map([$this, 'collectMeta'], $v);
        return $v;
    }
    private function resolveMerges($v, $depth = 0) {
        if ($depth > 20) throw new \Exception("Merge depth");
        if (is_array($v) && !array_is_list($v)) {
            if (isset($v["<<"])) {
                $merges = is_array($v["<<"]) && array_is_list($v["<<"]) ? $v["<<"] : [$v["<<"]]; unset($v["<<"]);
                foreach ($merges as $m) {
                    $rm = $m; if (is_string($rm) && $rm[0] === '*') $rm = $this->anchors[substr($rm, 1)] ?? null;
                    $rm = $this->resolveMerges($rm, $depth + 1);
                    if (is_array($rm)) { foreach ($rm as $mk => $mv) if ($mk[0] !== '_' && !array_key_exists($mk, $v)) $v[$mk] = $mv; }
                }
            }
            $res = []; foreach ($v as $k => $val) $res[$k] = $this->resolveMerges($val, $depth); return $res;
        }
        if (is_array($v)) return array_map(function($x) use ($depth) { return $this->resolveMerges($x, $depth); }, $v);
        return $v;
    }
    private function resolveEnv($v) {
        if (is_array($v) && !array_is_list($v)) {
            if (count($v) === 1 && isset($v["env"]) && is_string($v["env"])) {
                $var = $v["env"]; if (empty($this->app_env) && empty($this->doc_env) || isset($this->app_env[$var]) || isset($this->doc_env[$var])) return getenv($var) ?: "";
            }
            $res = []; foreach ($v as $k => $val) $res[$k] = $this->resolveEnv($val); return $res;
        }
        if (is_array($v)) return array_map([$this, 'resolveEnv'], $v);
        return $v;
    }
    private function resolveInterp($v, $root, $depth = 0) {
        if ($depth > 50) throw new \Exception("Interp depth");
        if (is_string($v) && strpos($v, '${') !== false) {
            $res = ""; $i = 0;
            while ($i < strlen($v)) {
                if (substr($v, $i, 2) === '${') {
                    $end = strpos($v, '}', $i + 2); if ($end === false) break;
                    $path = substr($v, $i + 2, $end - $i - 2);
                    $curr = $root; foreach (explode('.', $path) as $p) { if (!is_array($curr) || !array_key_exists($p, $curr)) throw new \Exception("Undef: $p"); $curr = $curr[$p]; }
                    $vs = is_string($curr) ? $curr : $this->dump($curr);
                    if (strpos($vs, '${') !== false) $vs = $this->resolveInterp($vs, $root, $depth + 1);
                    $res .= $vs; $i = $end + 1;
                } else { $res .= $v[$i]; $i++; }
            }
            return $res;
        }
        if (is_array($v) && !array_is_list($v)) { $res = []; foreach ($v as $k => $val) $res[$k] = $this->resolveInterp($val, $root, $depth); return $res; }
        if (is_array($v)) return array_map(function($x) use ($root, $depth) { return $this->resolveInterp($x, $root, $depth); }, $v);
        return $v;
    }
    private function resolveCoerce($v) {
        if (is_array($v) && !array_is_list($v)) {
            if (isset($v["_coerce"])) {
                $c = $v["_coerce"]; $t = $c["type"]; $val = $this->resolveCoerce($c["value"]);
                $s = is_string($val) ? $val : $this->dump($val);
                if ($t === "int") return intval($s); if ($t === "float") return floatval($s); if ($t === "bool") return strtolower($s) === "true"; return $s;
            }
            $res = []; foreach ($v as $k => $val) $res[$k] = $this->resolveCoerce($val); return $res;
        }
        if (is_array($v)) return array_map([$this, 'resolveCoerce'], $v);
        return $v;
    }
}

function parse($text, $app_env = []) { return (new Evaluator($app_env))->evaluate((new Parser($text, true))->parse()); }

File: include/nole.rs
use std::collections::{BTreeMap, HashSet};
use std::time::{Instant, Duration};
use std::env;

#[derive(Debug, Clone, PartialEq)]
pub enum NolValue {
    Null, Bool(bool), Int(i64), Float(f64), String(String), Array(Vec<NolValue>), Object(BTreeMap<String, NolValue>),
}

impl NolValue {
    pub fn dump(&self, indent: usize, level: usize, is_root: bool) -> String {
        match self {
            NolValue::Null => "null".to_string(),
            NolValue::Bool(b) => b.to_string(),
            NolValue::Int(i) => i.to_string(),
            NolValue::Float(f) => { let mut s = format!("{:.5}", f); while s.ends_with('0') { s.pop(); } if s.ends_with('.') { s.push('0'); } s }
            NolValue::String(s) => if is_root { s.clone() } else { format!("{:?}", s) },
            NolValue::Array(a) => format!("[{}]", a.iter().map(|v| v.dump(indent, level + 1, false)).collect::<Vec<_>>().join(", ")),
            NolValue::Object(o) => {
                let pad = " ".repeat(level * indent); let next_pad = " ".repeat((level + 1) * indent);
                let mut lines = Vec::new();
                for (k, v) in o { if k.starts_with('_') && !["_env", "_interpolate", "_meta"].contains(&k.as_str()) { continue; } let val = v.dump(indent, level + 1, false); if is_root { lines.push(format!("{}: {}", k, val)); } else { lines.push(format!("\n{}{}: {}", next_pad, k, val)); } }
                if is_root { return lines.join("\n"); } if o.is_empty() { return "{}".to_string(); }
                format!("{{{}\n{}}}", lines.join(","), pad)
            }
        }
    }
}

pub struct NolParser<'a> { text: &'a str, pos: usize, start: Instant, depth: usize, nole: bool }
impl<'a> NolParser<'a> {
    pub fn new(text: &'a str, nole: bool) -> Self { Self { text, pos: 0, start: Instant::now(), depth: 0, nole } }
    fn peek(&self) -> Option<char> { self.text[self.pos..].chars().next() }
    fn advance(&mut self) -> Option<char> { let c = self.peek()?; self.pos += c.len_utf8(); Some(c) }
    fn skip_ws(&mut self) { while let Some(c) = self.peek() { if c.is_whitespace() { self.advance(); } else if c == '#' { self.advance(); if self.peek() == Some('#') { self.advance(); while self.pos < self.text.len() - 1 && &self.text[self.pos..self.pos+2] != "##" { self.advance(); } if self.pos < self.text.len() { self.advance(); self.advance(); } } else { while let Some(c) = self.peek() { if c == '\n' { break; } self.advance(); } } } else { break; } } }
    fn parse_hex(&mut self, n: usize) -> char { let mut h = String::new(); for _ in 0..n { h.push(self.advance().expect("Hex")); } std::char::from_u32(u32::from_str_radix(&h, 16).expect("HexVal")).expect("Char") }
    fn read_str(&mut self, q: char) -> String { let mut s = String::new(); while let Some(c) = self.peek() { if c == q { self.advance(); return s; } if c == '\\' { self.advance(); let e = self.advance().expect("Esc"); if e == 'u' { s.push(self.parse_hex(4)); } else if e == 'U' { s.push(self.parse_hex(8)); } else { s.push(match e { 'n'=>'\n', 'r'=>'\r', 't'=>'\t', x=>x }); } } else { s.push(self.advance().unwrap()); } } s }
    fn read_key(&mut self) -> String { self.skip_ws(); let c = self.peek().expect("Key"); if c == '"' || c == '\'' { let q = self.advance().unwrap(); self.read_str(q) } else { let mut s = String::new(); while let Some(c) = self.peek() { if c.is_alphanumeric() || c == '_' || c == '-' { s.push(self.advance().unwrap()); } else { break; } } s } }
    pub fn parse_value(&mut self) -> NolValue {
        self.skip_ws(); if self.start.elapsed() > Duration::from_secs(1) { panic!("Timeout"); } self.depth += 1; if self.depth > 100 { panic!("Depth"); }
        let c = self.peek().expect("EOF");
        let res = match c {
            '{' => { self.advance(); let mut o = BTreeMap::new(); while let Some(_) = self.peek() { self.skip_ws(); if self.peek() == Some('}') { self.advance(); break; } self.parse_pair(&mut o); self.skip_ws(); if self.peek() == Some(',') { self.advance(); } } NolValue::Object(o) }
            '[' => { self.advance(); let mut a = Vec::new(); while let Some(_) = self.peek() { self.skip_ws(); if self.peek() == Some(']') { self.advance(); break; } a.push(self.parse_value()); self.skip_ws(); if self.peek() == Some(',') { self.advance(); } } NolValue::Array(a) }
            '"' | '\'' => { let q = self.advance().unwrap(); NolValue::String(self.read_str(q)) }
            '&' | '*' | '<' if self.nole => {
                let op = self.advance().unwrap();
                if op == '<' { if self.peek() == Some('<') { self.pos -= 1; self.depth -= 1; return NolValue::Null; } let mut t = String::new(); while let Some(c) = self.peek() { if c == '>' { break; } t.push(self.advance().unwrap()); } if self.peek() == Some('>') { self.advance(); } let val = self.parse_value(); let mut ci = BTreeMap::new(); ci.insert("type".to_string(), NolValue::String(t)); ci.insert("value".to_string(), val); let mut co = BTreeMap::new(); co.insert("_coerce".to_string(), NolValue::Object(ci)); NolValue::Object(co) }
                else if op == '*' { let mut n = String::new(); while let Some(c) = self.peek() { if c.is_alphanumeric() || c == '_' || c == '-' { n.push(self.advance().unwrap()); } else { break; } } NolValue::String(format!("*{}", n)) }
                else { let mut n = String::new(); while let Some(c) = self.peek() { if c.is_alphanumeric() || c == '_' || c == '-' { n.push(self.advance().unwrap()); } else { break; } } let val = self.parse_value(); let mut ai = BTreeMap::new(); ai.insert("name".to_string(), NolValue::String(n)); ai.insert("value".to_string(), val); let mut an = BTreeMap::new(); an.insert("_anchor".to_string(), NolValue::Object(ai)); NolValue::Object(an) }
            }
            '0'..='9' | '-' => { let mut s = String::new(); while let Some(c) = self.peek() { if c.is_ascii_digit() || ".-eE+".contains(c) { s.push(self.advance().unwrap()); } else { break; } } if s.contains('.') || s.to_lowercase().contains('e') { NolValue::Float(s.parse().unwrap()) } else { NolValue::Int(s.parse().unwrap()) } }
            _ => { let mut s = String::new(); while let Some(c) = self.peek() { if c.is_alphabetic() { s.push(self.advance().unwrap()); } else { break; } } match s.as_str() { "true" => NolValue::Bool(true), "false" => NolValue::Bool(false), "null" => NolValue::Null, _ => panic!("Invalid: {}", s) } }
        }; self.depth -= 1; res
    }
    fn parse_pair(&mut self, o: &mut BTreeMap<String, NolValue>) {
        self.skip_ws(); if self.nole && self.peek() == Some('&') {
            self.advance(); let n = self.read_key(); self.skip_ws(); let mut ai = BTreeMap::new(); ai.insert("name".to_string(), NolValue::String(n));
            if self.peek() == Some(':') { self.advance(); ai.insert("value".to_string(), self.parse_value()); } else { ai.insert("value".to_string(), NolValue::Null); }
            if !o.contains_key("_anchors") { o.insert("_anchors".to_string(), NolValue::Array(Vec::new())); }
            if let Some(NolValue::Array(a)) = o.get_mut("_anchors") { a.push(NolValue::Object(ai)); } return;
        }
        let mut is_m = false; if self.nole && self.peek() == Some('<') { self.advance(); if self.peek() == Some('<') { self.advance(); is_m = true; } else { self.pos -= 1; } }
        let key = if is_m { "<<".to_string() } else { self.read_key() };
        if !self.nole && ["_env", "_interpolate", "_meta"].contains(&key.as_str()) { panic!("Reserved: {}", key); }
        self.skip_ws(); if self.advance() != Some(':') { panic!("Expected : for {}", key); }
        let val = self.parse_value(); if is_m { if !o.contains_key("<<") { o.insert("<<".to_string(), NolValue::Array(Vec::new())); } if let Some(NolValue::Array(a)) = o.get_mut("<<") { a.push(val); } } else { if o.contains_key(&key) { panic!("Duplicate: {}", key); } o.insert(key, val); }
    }
    pub fn parse(&mut self) -> BTreeMap<String, NolValue> {
        let mut root = BTreeMap::new();
        while self.pos < self.text.len() {
            self.skip_ws(); if self.peek().is_none() { break; }
            if self.peek() == Some('[') {
                self.advance(); let mut is_a = false; if self.peek() == Some('*') { self.advance(); is_a = true; }
                let mut path = String::new(); while let Some(c) = self.peek() { if c == ']' { self.advance(); break; } path.push(self.advance().unwrap()); }
                let parts: Vec<String> = path.split('.').map(|s| s.to_string()).collect(); let mut curr = &mut root;
                for (i, p) in parts.iter().enumerate() {
                    let last = i == parts.len() - 1; if !curr.contains_key(p) { curr.insert(p.clone(), if last && is_a { NolValue::Array(Vec::new()) } else { NolValue::Object(BTreeMap::new()) }); }
                    if last {
                        if is_a { if let Some(NolValue::Array(a)) = curr.get_mut(p) { a.push(NolValue::Object(BTreeMap::new())); let idx = a.len() - 1; if let Some(NolValue::Object(o)) = a.get_mut(idx) { self.parse_into(o); } } }
                        else if let Some(NolValue::Object(o)) = curr.get_mut(p) { self.parse_into(o); }
                    } else if let Some(NolValue::Object(o)) = curr.get_mut(p) { curr = o; } else { panic!("Collision"); }
                }
            } else { self.parse_pair(&mut root); }
        } root
    }
    fn parse_into(&mut self, o: &mut BTreeMap<String, NolValue>) { while self.pos < self.text.len() { self.skip_ws(); match self.peek() { Some('[') | None => break, _ => self.parse_pair(o) } } }
}

pub struct Evaluator { anchors: BTreeMap<String, NolValue>, app_env: HashSet<String>, doc_env: HashSet<String> }
impl Evaluator {
    pub fn new(app_env: Vec<String>) -> Self { Self { anchors: BTreeMap::new(), app_env: app_env.into_iter().collect(), doc_env: HashSet::new() } }
    pub fn evaluate(&mut self, mut root: NolValue) -> NolValue { root = self.collect_meta(root); root = self.resolve_merges(root, 0); root = self.resolve_env(root); let root_clone = root.clone(); root = self.resolve_interp(root, &root_clone, 0); self.resolve_coerce(root) }
    fn collect_meta(&mut self, v: NolValue) -> NolValue {
        match v {
            NolValue::Object(mut o) => {
                if let Some(NolValue::Array(ans)) = o.remove("_anchors") { for av in ans { if let NolValue::Object(mut a) = av { if let Some(NolValue::String(n)) = a.remove("name") { let val_raw = a.remove("value").unwrap(); let val = if val_raw == NolValue::Null { NolValue::Object(o.clone()) } else { val_raw }; let val = self.collect_meta(val); self.anchors.insert(n, val); } } } }
                if let Some(NolValue::Object(mut a)) = o.remove("_anchor") { let n = if let Some(NolValue::String(n)) = a.remove("name") { n } else { panic!() }; let val_raw = a.remove("value").unwrap(); let val = if val_raw == NolValue::Null { NolValue::Object(o.clone()) } else { val_raw }; let val = self.collect_meta(val); self.anchors.insert(n, val.clone()); return val; }
                if let Some(NolValue::Object(e)) = o.get("_env") { if let Some(NolValue::Array(al)) = e.get("allowed") { for x in al { if let NolValue::String(s) = x { self.doc_env.insert(s.clone()); } } } }
                NolValue::Object(o.into_iter().map(|(k, val)| (k, self.collect_meta(val))).collect())
            }
            NolValue::Array(a) => NolValue::Array(a.into_iter().map(|x| self.collect_meta(x)).collect()), _ => v
        }
    }
    fn resolve_merges(&mut self, v: NolValue, depth: usize) -> NolValue {
        if depth > 20 { panic!("Merge depth"); }
        match v {
            NolValue::Object(mut o) => {
                if let Some(merges) = o.remove("<<") { let ma = match merges { NolValue::Array(a) => a, _ => vec![merges] }; for m in ma { let mut rm = m; if let NolValue::String(ref s) = rm { if s.starts_with('*') { let an = &s[1..]; rm = self.anchors.get(an).expect("Undef anchor").clone(); } } rm = self.resolve_merges(rm, depth + 1); if let NolValue::Object(mo) = rm { for (mk, mv) in mo { if mk.starts_with('_') { continue; } if !o.contains_key(&mk) { o.insert(mk, mv); } } } } }
                NolValue::Object(o.into_iter().map(|(k, val)| (k, self.resolve_merges(val, depth))).collect())
            }
            NolValue::Array(a) => NolValue::Array(a.into_iter().map(|x| self.resolve_merges(x, depth)).collect()), _ => v
        }
    }
    fn resolve_env(&mut self, v: NolValue) -> NolValue {
        match v {
            NolValue::Object(o) => { if o.len() == 1 && o.contains_key("env") { if let Some(NolValue::String(var)) = o.get("env") { if self.app_env.is_empty() && self.doc_env.is_empty() || self.app_env.contains(var) || self.doc_env.contains(var) { let val = env::var(var).unwrap_or_default(); if val.len() > 65536 { panic!("Env size limit"); } return NolValue::String(val); } panic!("Env denied: {}", var); } } NolValue::Object(o.into_iter().map(|(k, val)| (k, self.resolve_env(val))).collect()) }
            NolValue::Array(a) => NolValue::Array(a.into_iter().map(|x| self.resolve_env(x)).collect()), _ => v
        }
    }
    fn resolve_interp(&mut self, v: NolValue, root: &NolValue, depth: usize) -> NolValue {
        if depth > 50 { panic!("Interp depth"); }
        match v {
            NolValue::String(s) if s.contains("${") => {
                let mut res = String::new(); let mut i = 0; let b = s.as_bytes();
                while i < b.len() { if b[i] == b'$' && i + 1 < b.len() && b[i+1] == b'{' { let start = i + 2; let mut end = start; while end < b.len() && b[end] != b'}' { end += 1; } if end == b.len() { panic!("Unclosed"); } let path = &s[start..end]; let mut curr = root; for p in path.split('.') { if let NolValue::Object(o) = curr { curr = o.get(p).expect("Undef interp"); } else { panic!("Undef interp"); } } let mut vs = match curr { NolValue::String(ref s) => s.clone(), _ => curr.dump(2, 0, true) }; if vs.contains("${") { vs = match self.resolve_interp(NolValue::String(vs), root, depth + 1) { NolValue::String(s) => s, _ => panic!() }; } res.push_str(&vs); i = end + 1; } else { res.push(b[i] as char); i += 1; } }
                if res.len() > 10*1024*1024 { panic!("Interp size limit"); }
                NolValue::String(res)
            }
            NolValue::Object(o) => NolValue::Object(o.into_iter().map(|(k, val)| (k, self.resolve_interp(val, root, depth))).collect()),
            NolValue::Array(a) => NolValue::Array(a.into_iter().map(|x| self.resolve_interp(x, root, depth)).collect()), _ => v
        }
    }
    pub fn resolve_coerce(&mut self, v: NolValue) -> NolValue {
        match v {
            NolValue::Object(mut o) => { if let Some(NolValue::Object(mut c)) = o.remove("_coerce") { if let Some(NolValue::String(t)) = c.remove("type") { let val_raw = c.remove("value").unwrap(); let val = self.resolve_coerce(val_raw); let s = match val { NolValue::String(ref s) => s.clone(), _ => val.dump(2, 0, false) }; return match t.as_str() { "int" => NolValue::Int(s.parse().unwrap()), "float" => NolValue::Float(s.parse().unwrap()), "bool" => NolValue::Bool(s.to_lowercase() == "true"), _ => NolValue::String(s) }; } } NolValue::Object(o.into_iter().map(|(k, val)| (k, self.resolve_coerce(val))).collect()) }
            NolValue::Array(a) => NolValue::Array(a.into_iter().map(|x| self.resolve_coerce(x)).collect()), _ => v
        }
    }
}
pub fn parse(text: &str, app_env: Vec<String>) -> NolValue { let mut p = NolParser::new(text, true); let root = NolValue::Object(p.parse()); Evaluator::new(app_env).evaluate(root) }
